Add: log.Flush()

matthewmueller opened the pull request:
This PR gets the `log.Flush()` work started. It currently just demonstrates the flow and the API, not the implementation.

Sorry, I hate receiving WIP PRs because they usually just hang forever, but I'm definitely going to get this one finished up ASAP. 

Could just use a quick review to make sure this is on the right track with what you're looking for, mainly around adding `log.Flush()` to the top-level API. At first I thought this would be a breaking change, but now I'm not so sure, since I didn't alter the interface at all. 

The rest of the change is just setting up some testing for papertrail and also firehose when I implement it :-)

tj commented:
woot awesome, thanks for working on this! Direction looks good to me, maybe worth chucking it in ./internal so we don't have rando's using it as public API (unless you want to move buffer to a public repo maybe)

matthewmueller commented:
Alrighty there's a few todos in there, but overall this one should be good to go!

It ended up being slightly different than what I had before. The buffer was doing some of the heavy lifting that TCP already does and wasn't doing a very good job of it + the API was a bit wonky. I replaced that with a queue that drops logs when it's at capacity. 

Here's what's changed from master:

- Added log.Flush() to the API
- All papertrail traffic is done over TCP + TLS
- Flushing will block until the queue is fully drained, making it a much better fit for lambda functions

The queue is in `internal/queue`. Even though it's fairly generic, I opted for this as a opposed to a separate package because it's not really special in Go (can be easily written), but should be useful in some of the other handlers.

tj commented on handlers/papertrail/papertrail.go:
```diff
@@ -26,50 +28,136 @@ type Config struct {
+	// TODO: see if papertrail can handle out of order
+	// logs. if so, we can make our write function
+	// thread-safe & adjust the concurrency here
+	q := queue.New(config.Capacity, 1)
```
Ahh I was thinking we'd have a `Buffer` with the size/interval flushing. The channel stuff is basically a queue as-is but the flushing will definitely be common among them.

As far as dropping when the threshold is reached, yea I guess we just delegate and let the user deal with it ahah not sure, definitely a tricky situation hahah. `stdlog`'ing it is probably good at least, so they show up in local logs at least (or cloudwatch in the Lambda cas

tj replied:
I suppose concurrency doesn't really matter either, if the Loggly and friends are rate-limiting or congested the concurrency wouldn't really help anyway.

matthewmueller replied:
That's what I originally had, but after digging into TCP a bit more, it didn't really makes sense to buffer with an open TCP connection, TCP will already do that and probably do a better job than us.

Yep, the queue is basically just a channel, it just adds support for: concurrency, knowing when it's drained and dropping when it's full. For things like Firehose or Cloudwatch, you can issue HTTP requests concurrently, which would hopefully speed things up and if you have a really fast writer.

This TODO is just to see if we could have multiple goroutines writing on the TCP connection at once and Papertrail being able to rearrange the logs in the console based on the timestamp. Not too important though, since I think Papertrail isn't even granular enough for this to matter.

Oh good call on logging out when we're at capacity, I'm printing the error, but I should also write out the actual log. Anything else you'd like me to change / fix?

tj replied:
Ohh for TCP yeah true, I was thinking for HTTP like Elasticsearch / Loggly etc where it's a lot more efficient to batch. The HTTP envelope alone is usually bigger than a log line so I think it's worth it there

matthewmueller replied:
Ah, I see yah. I can add that original buffer back in then for us in some of the other providers? Ideally it would probably issue requests as fast as the reader can handle, but then start batching backed up logs. That might get tricky though haha.

matthewmueller commented:
closing. while it might be nice to switch papertrail over to TCP (from UDP), you can add an logger that has a flush method that also implements log.Interface in your own code. Here's an example: https://github.com/matthewmueller/go-datadog

This doesn't need to be in here.
